File: backend/src/nvidia.js
const NVIDIA_URL = "https://integrate.api.nvidia.com/v1/chat/completions";

function extractText(data) {
  const choice = data?.choices?.[0];
  return (
    choice?.message?.content ||
    choice?.delta?.content ||
    "No response returned by the language model."
  );
}

export async function generateWithNvidia({ apiKey, model, system, prompt }) {
  const response = await fetch(NVIDIA_URL, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${apiKey}`
    },
    body: JSON.stringify({
      model,
      messages: [
        { role: "system", content: system },
        { role: "user", content: prompt }
      ],
      temperature: 0.6,
      top_p: 0.7,
      max_tokens: 4096,
      stream: false
    })
  });

  if (!response.ok) {
    const message = await response.text();
    throw new Error(`NVIDIA API error ${response.status}: ${message}`);
  }

  const data = await response.json();
  return extractText(data);
}

File: backend/internal/app/nvidia.go
package app

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"time"
)

const ollamaCloudURL = "https://api.ollama.com/v1/chat/completions"

type aiClient struct {
	httpClient *http.Client
	apiKey     string
	model      string
	baseURL    string
}

func newAIClient(apiKey, model, baseURL string) *aiClient {
	if baseURL == "" {
		baseURL = ollamaCloudURL
	}
	return &aiClient{
		apiKey:  apiKey,
		model:   model,
		baseURL: baseURL,
		httpClient: &http.Client{
			Timeout: 45 * time.Second,
		},
	}
}

func (c *aiClient) generate(ctx context.Context, system, prompt string) (string, error) {
	payload := map[string]interface{}{
		"model": c.model,
		"messages": []map[string]string{
			{"role": "system", "content": system},
			{"role": "user", "content": prompt},
		},
		"temperature": 0.6,
		"top_p":       0.7,
		"max_tokens":  4096,
		"stream":      false,
	}

	body, err := json.Marshal(payload)
	if err != nil {
		return "", err
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.baseURL, bytes.NewReader(body))
	if err != nil {
		return "", err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+c.apiKey)

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	raw, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}
	if resp.StatusCode >= 400 {
		return "", fmt.Errorf("AI API error %d: %s", resp.StatusCode, string(raw))
	}

	var parsed struct {
		Choices []struct {
			Message struct {
				Content string `json:"content"`
			} `json:"message"`
		} `json:"choices"`
	}
	if err := json.Unmarshal(raw, &parsed); err != nil {
		return "", err
	}
	if len(parsed.Choices) == 0 {
		return "No response returned by the language model.", nil
	}

	return parsed.Choices[0].Message.Content, nil
}
